rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java

```diff
@@ -206,23 +250,30 @@ private boolean allowByMinerNotBanned(Peer sender, Message message) {
      * record event if message is repeated
      */
     private boolean allowByMessageUniqueness(Peer sender, Message message) {
-        Keccak256 encodedMessage = new Keccak256(HashUtil.keccak256(message.getEncoded()));
+        if (message.getMessageType() != MessageType.BLOCK_MESSAGE && message.getMessageType() != MessageType.TRANSACTIONS) {
+            return true;
+        }
 
-        boolean contains = receivedMessages.contains(encodedMessage);
+        long currentTime = System.currentTimeMillis();
 
-        if (!contains) {
-            if (message.getMessageType() == MessageType.BLOCK_MESSAGE || message.getMessageType() == MessageType.TRANSACTIONS) {
-                if (this.receivedMessages.size() >= MAX_NUMBER_OF_MESSAGES_CACHED) {
-                    this.receivedMessages.clear();
-                }
-                this.receivedMessages.add(encodedMessage);
-            }
+        synchronized (receivedPeerMessages) {
```

Comment:
what do you think about placing this cleanup logic between the `if (receivedPeerMessages.containsKey() {...` check and `this.receivedPeerMessages.put(receivedPeerMessageKey, currentTime);`? In this case some useful entries may still be available for the check before the cleanup.

also you could place entire work with the `receivedPeerMessages` collection inside the `synchronized` block, so that in this case there's no need to wrap the map with `Collections.synchronizedMap()` and do double synchronisation.

Reply:
Addressed